File: WEASEL/Python/weasel.py
import random
import time
import copy
from string import ascii_uppercase

TARGET = list("METHINKS IT IS LIKE A WEASEL")
NUM_OFFSPRINGS = 100
MAX_SCORE = 28

class Child:
    def __init__(self, string = []):
        self.str = string
        self.score = len(list(filter(lambda x: x[0] == x[1], zip(self.str, TARGET))))
    
    def __str__(self):
       return (str.format("{} - SCORE: {}".format("".join(current_child.str), current_child.score)))

current_child = Child()
childs = []

def get_random_str():
    rand_str = []
    for i in range(len(TARGET)):
        rand_str.append(random.choice(ascii_uppercase + ' '))
    return rand_str

def evaluate(input):
    score = 0
    for i in range(len(TARGET)):
        if(input.str[i] == TARGET[i]):
            score += 1
    input.score = score

def mutate():
    global childs
    for i in (childs):
        for j in range(len(TARGET)):
            if(random.randint(0, 100) <= 5):
                i.str[j] = random.choice(ascii_uppercase + ' ')
        evaluate(i)
            
def reproduce():
    global childs
    global current_child
    childs = []
    for i in range(NUM_OFFSPRINGS):
        child = Child(current_child.str)
        childs.append(copy.deepcopy(child))

def evolve():
    global childs
    global current_child
    for i in childs:
        if(i.score > current_child.score):
            current_child = i

def start_program():
    global current_child
    current_child = Child(get_random_str())
    print(current_child)

    while(current_child.score < MAX_SCORE):
        time.sleep(0.1)

        reproduce()
        mutate()
        evolve()

        print(current_child)

start_program()

File: WEASEL/C/weasel.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#define NUM_OFFSPRINGS 100
#define MAX_SCORE 28

char target[28] = {"METHINKS IT IS LIKE A WEASEL"};

typedef struct child
{
    char text[28];
    int score;
}
child;

child current;

void generateRandomString(child *input)
{
    for(int i = 0; i < 28; i++)
    {
        input->text[i] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ "[random () % 27];
    }
}

void evaluate(child *input)
{
    int score = 0;
    for(int i = 0; i < 28; i++)
    {
        if((input->text[i] == target[i]) == 1)
        {
            score++;
        }
    }
    input->score = score;
}

void reproduce(child *childs)
{
    for(int i = 0; i < NUM_OFFSPRINGS; i++)
    {
        childs[i] = current;
    }
}

void mutate(child *childs)
{
    for(int i = 0; i < NUM_OFFSPRINGS; i++)
    {
        for(int j = 0; j < 28; j++)
        {
            if(rand() % 100 < 5)
            {
                childs[i].text[j] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ "[random () % 27];
            }
        }
        evaluate(&childs[i]);
    }
}

void evolve(child *childs)
{
    for(int i = 0; i < NUM_OFFSPRINGS; i++)
    {
        if(childs[i].score > current.score)
        {
            current = childs[i];
        }
    }
}

void main()
{
    srand(time(NULL));
    generateRandomString(&current);
    evaluate(&current);
    child childs[100];

    while(current.score < MAX_SCORE)
    {
        reproduce(childs);
        mutate(childs);
        evolve(childs);
        for(int i = 0; i < 28; i++)
        {
            printf("%c", current.text[i]);
        }
        printf(" - SCORE: %d\n", current.score);
    }

}